Here is another test that should not fail:
```
from sympy.physics.units import Quantity, Dimension, length, time, meter, second
from sympy import Abs
v_w1 = Quantity('v_w1', length/time, 1.5*meter/second)
expr = 2.5*second/meter*v_w1 - 3000
Quantity._collect_factor_and_dimension(expr)

...
--> 123                 assert dim == addend_dim
    124                 factor += addend_factor
    125             return factor, dim

AssertionError: 
```
I believe that this is because the dimension of `3000` is `1` instead of `Dimension(1)`:
```
Quantity._collect_factor_and_dimension(3000)

(3000, 1)
```

--- a/sympy/physics/units/quantities.py
+++ b/sympy/physics/units/quantities.py
@@ -6,10 +6,10 @@ Physical quantities.
 
 from __future__ import division
 
+from sympy import Add, AtomicExpr, Basic, Function, Mul, Pow, S, Symbol, \
+    sympify
 from sympy.core.compatibility import string_types
-from sympy import sympify, Mul, Pow, S, Symbol, Add, AtomicExpr, Basic
-from sympy.physics.units import Dimension
-from sympy.physics.units import dimensions
+from sympy.physics.units import Dimension, dimensions
 from sympy.physics.units.prefixes import Prefix
 
 
@@ -123,6 +123,9 @@ class Quantity(AtomicExpr):
                 assert dim == addend_dim
                 factor += addend_factor
             return factor, dim
+        elif isinstance(expr, Function):
+            fds = [Quantity._collect_factor_and_dimension(arg) for arg in expr.args]
+            return expr.func(*(f[0] for f in fds)), expr.func(*(d[1] for d in fds))
         else:
             return expr, 1
 

--- a/sympy/physics/units/quantities.py
+++ b/sympy/physics/units/quantities.py
@@ -102,11 +102,12 @@ class Quantity(AtomicExpr):
 
     @staticmethod
     def _collect_factor_and_dimension(expr):
+        """Return tuple with factor expression and dimension expression."""
         if isinstance(expr, Quantity):
             return expr.scale_factor, expr.dimension
         elif isinstance(expr, Mul):
             factor = 1
-            dimension = 1
+            dimension = Dimension(1)
             for arg in expr.args:
                 arg_factor, arg_dim = Quantity._collect_factor_and_dimension(arg)
                 factor *= arg_factor
@@ -127,7 +128,7 @@ class Quantity(AtomicExpr):
             fds = [Quantity._collect_factor_and_dimension(arg) for arg in expr.args]
             return expr.func(*(f[0] for f in fds)), expr.func(*(d[1] for d in fds))
         else:
-            return expr, 1
+            return expr, Dimension(1)
 
     def convert_to(self, other):
         """

--- a/sympy/physics/units/tests/test_quantities.py
+++ b/sympy/physics/units/tests/test_quantities.py
@@ -2,14 +2,16 @@
 
 from __future__ import division
 
-from sympy import Symbol, Add, Number, S, integrate, sqrt, Rational, Abs, diff, symbols, Basic
-from sympy.physics.units import convert_to, find_unit
-
-from sympy.physics.units.definitions import s, m, kg, speed_of_light, day, minute, km, foot, meter, grams, amu, au, \
-    quart, inch, coulomb, millimeter, steradian, second, mile, centimeter, hour
-from sympy.physics.units.dimensions import length, time, charge
-from sympy.physics.units.quantities import Quantity
+from sympy import Abs, Add, Basic, Number, Rational, S, Symbol, diff, exp, \
+    integrate, log, sqrt, symbols
+from sympy.physics.units import convert_to, find_unit, amount_of_substance, \
+    volume
+from sympy.physics.units.definitions import amu, au, centimeter, coulomb, day, \
+    foot, grams, hour, inch, kg, km, m, meter, mile, millimeter, minute, mole, \
+    quart, s, second, speed_of_light, steradian
+from sympy.physics.units.dimensions import charge, length, time
 from sympy.physics.units.prefixes import PREFIXES, kilo
+from sympy.physics.units.quantities import Quantity
 from sympy.utilities.pytest import raises
 
 k = PREFIXES["k"]
@@ -246,3 +248,18 @@ def test_factor_and_dimension():
     x, y = symbols('x y')
     assert (x + y/100, length) == Quantity._collect_factor_and_dimension(
         x*m + y*centimeter)
+
+    cH = Quantity('cH', amount_of_substance/volume)
+    pH = -log(cH)
+
+    assert (1, volume/amount_of_substance) == Quantity._collect_factor_and_dimension(
+        exp(pH))
+
+    v_w1 = Quantity('v_w1', length/time, 1.5*meter/second)
+    v_w2 = Quantity('v_w2', length/time, 2*meter/second)
+    expr = Abs(v_w1/2 - v_w2)
+    assert (1.25000000000000, length/time) == \
+        Quantity._collect_factor_and_dimension(expr)
+
+    expr = v_w1 - Abs(v_w1)
+    assert (0, 1) == Quantity._collect_factor_and_dimension(expr)

--- a/sympy/physics/units/tests/test_quantities.py
+++ b/sympy/physics/units/tests/test_quantities.py
@@ -9,7 +9,7 @@ from sympy.physics.units import convert_to, find_unit, amount_of_substance, \
 from sympy.physics.units.definitions import amu, au, centimeter, coulomb, day, \
     foot, grams, hour, inch, kg, km, m, meter, mile, millimeter, minute, mole, \
     quart, s, second, speed_of_light, steradian
-from sympy.physics.units.dimensions import charge, length, time
+from sympy.physics.units.dimensions import Dimension, charge, length, time
 from sympy.physics.units.prefixes import PREFIXES, kilo
 from sympy.physics.units.quantities import Quantity
 from sympy.utilities.pytest import raises
@@ -241,6 +241,7 @@ def test_sum_of_incompatible_quantities():
 
 
 def test_factor_and_dimension():
+    assert (3000, Dimension(1)) == Quantity._collect_factor_and_dimension(3000)
     assert (1001, length) == Quantity._collect_factor_and_dimension(meter + km)
     assert (2, length/time) == Quantity._collect_factor_and_dimension(
         meter/second + 36*km/(10*hour))
@@ -262,4 +263,8 @@ def test_factor_and_dimension():
         Quantity._collect_factor_and_dimension(expr)
 
     expr = v_w1 - Abs(v_w1)
-    assert (0, 1) == Quantity._collect_factor_and_dimension(expr)
+    assert (0, Dimension(1)) == Quantity._collect_factor_and_dimension(expr)
+
+    expr = 2.5*second/meter*v_w1 - 3000
+    assert (-2996.25000000000, Dimension(1)) == \
+        Quantity._collect_factor_and_dimension(expr)

--- a/sympy/physics/units/tests/test_quantities.py
+++ b/sympy/physics/units/tests/test_quantities.py
@@ -12,7 +12,7 @@ from sympy.physics.units.definitions import amu, au, centimeter, coulomb, day, \
 from sympy.physics.units.dimensions import Dimension, charge, length, time
 from sympy.physics.units.prefixes import PREFIXES, kilo
 from sympy.physics.units.quantities import Quantity
-from sympy.utilities.pytest import raises
+from sympy.utilities.pytest import XFAIL, raises
 
 k = PREFIXES["k"]
 
@@ -256,15 +256,19 @@ def test_factor_and_dimension():
     assert (1, volume/amount_of_substance) == Quantity._collect_factor_and_dimension(
         exp(pH))
 
-    v_w1 = Quantity('v_w1', length/time, 1.5*meter/second)
+    v_w1 = Quantity('v_w1', length/time, S(3)/2*meter/second)
     v_w2 = Quantity('v_w2', length/time, 2*meter/second)
     expr = Abs(v_w1/2 - v_w2)
-    assert (1.25000000000000, length/time) == \
+    assert (S(5)/4, length/time) == \
         Quantity._collect_factor_and_dimension(expr)
 
+    expr = S(5)/2*second/meter*v_w1 - 3000
+    assert (-(2996 + S(1)/4), Dimension(1)) == \
+        Quantity._collect_factor_and_dimension(expr)
+
+
+@XFAIL
+def test_factor_and_dimension_with_Abs():
+    v_w1 = Quantity('v_w1', length/time, S(3)/2*meter/second)
     expr = v_w1 - Abs(v_w1)
-    assert (0, Dimension(1)) == Quantity._collect_factor_and_dimension(expr)
-
-    expr = 2.5*second/meter*v_w1 - 3000
-    assert (-2996.25000000000, Dimension(1)) == \
-        Quantity._collect_factor_and_dimension(expr)
+    assert (0, lenth/time) == Quantity._collect_factor_and_dimension(expr)

